sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -462,30 +630,104 @@ template <class KernelType, class KernelArgType, int DIMS> class libCMBatch {
     GroupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
     GroupDim[2] = (uint32_t)(GlobalSize[2] / LocalSize[2]);
 
-    PLACEHOLDER_UNUSED(WrappedLambda_NDITEM_3DIM);
-    // TODO : Invoke invokeLambda_NDITEM_3DIM through CM's multi-threaded
-    // kernel launching with WrappedLambda_NDITEM_3DIM and dimension info
+    ESimdCPUKernel ESimdCPU((fptrVoid)InvokeLambda_NDITEM_3DIM, GroupDim,
+                            SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_3DIM),
+                      WrappedLambda_NDITEM_3DIM.get());
   }
 };
 
+/// Implementation for ESIMD_CPU device interface accessing ESIMD
+/// intrinsics and LibCM functionalties requred by intrinsics
 // Intrinsics
 sycl::detail::ESIMDDeviceInterface::ESIMDDeviceInterface() {
-  reserved = nullptr;
   version = ESIMDEmuPluginInterfaceVersion;
+  reserved = nullptr;
 
-  /// TODO : Fill *_ptr fields with function pointers from CM
-  /// functions prefixed with 'cm_support'
+  /* From 'esimd_emu_functions_v1.h' : Start */
+  cm_barrier_ptr = cm_support::barrier;
+  cm_sbarrier_ptr = cm_support::split_barrier;
+  cm_fence_ptr = cm_support::fence;
 
-  cm_barrier_ptr = nullptr;  /* cm_support::barrier; */
-  cm_sbarrier_ptr = nullptr; /* cm_support::split_barrier; */
-  cm_fence_ptr = nullptr;    /* cm_support::fence; */
+  sycl_get_surface_base_addr_ptr = cm_support::get_surface_base_addr;
+  __cm_emu_get_slm_ptr = cm_support::get_slm_base;
+  cm_slm_init_ptr = cm_support::init_slm;
+  /* From 'esimd_emu_functions_v1.h' : End */
+}
+
+/// Implementation for Host Kernel Launch used by
+/// piEnqueueKernelLaunch
+template <typename T> using KernelFunc = std::function<void(T)>;
 
-  sycl_get_surface_base_addr_ptr =
-      nullptr;                    /* cm_support::get_surface_base_addr; */
-  __cm_emu_get_slm_ptr = nullptr; /* cm_support::get_slm_base; */
-  cm_slm_init_ptr = nullptr;      /* cm_support::init_slm; */
+template <int NDims> struct InvokeBaseImpl {
+  static sycl::range<NDims> get_range(const size_t *GlobalWorkSize);
+};
+
+static constexpr bool isNull(int NDims, const size_t *R) {
+  return ((0 == R[0]) && (1 > NDims || 0 == R[1]) && (2 > NDims || 0 == R[2]));
```

Comment:
say, NDims is 1, and R[0] is 0. Then an attempt to calculate `(1 > NDims || 0 == R[1])` will be made. Since `1 > NDims` is `false`, `R[1]` will be evaluated and cause undefined behavior.

Reply:
My comment was suggesting to fix a bug in the original line by changing `1 >= NDims` to `1 > NDims`. But it turns out that my suggestion doesn't fix the line. I think the correct one would be:
`return ((0 == R[0]) && (NDims < 2 || 0 == R[1]) && (NDims < 3 || 0 == R[2]))`

Reply:
I will apply @romanovvlad 's suggestion here.

Reply:
Suggestion is applied.